_posts/2017-02-02-Rust-1.15.md

```diff
@@ -0,0 +1,273 @@
+struct Pet {
+    name: String,
+}
+```
+
+The `Debug` trait would be implemented for `Pet`. However, this only worked for
+traits provided as part of the standard library; it was not customizable. With
+Rust 1.15, it now is. That means, if you want to turn your `Pet` into JSON, it's
+as easy as adding [serde] to your `Cargo.toml`:
+
+```toml
+[dependencies]
+serde = "0.9.5"
+serde_derive = "0.9.5"
+serde_json = "0.9.3"
+```
+
+[serde]: https://serde.rs
+
+And adding another trait to your `Pet`:
+
+```rust
+#[macro_use]
+extern crate serde_derive;
+
+extern crate serde_json;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Pet {
+    name: String,
+}
+
+fn main() {
+    let pet = Pet { name: String::from("Ferris") };
+
+    let serialized = serde_json::to_string(&pet).unwrap();
+    println!("serialized = {}", serialized);
+
+    let deserialized: Pet = serde_json::from_str(&serialized).unwrap();
+    println!("deserialized = {:?}", deserialized);
+}
+```
+
+This will output:
+
+```text
+serialized = {"name":"Ferris"}
+deserialized = Pet { name: "Ferris" }
+```
+
+Another common use-case is [Diesel](http://diesel.rs/). Say we wanted to store
+our `Pet`s in a database. We could do that like this:
+
+```rust
+// some extern crate and use lines elided here
+
+#[derive(Queryable)]
+strut Pet {
+    name: String,
+}
+
+fn main() {
+    use diesel_demo::schema::pets::dsl::*;
+
+    let connection = establish_connection();
+    let results = pets 
+        .limit(5)
+        .load::<Pet>(&connection)
+        .expect("Error loading pets");
+
+    println!("Displaying {} pets", results.len());
+    for pet in results {
+        println!("{}", pet.name);
+    }
+}
+```
+
+For full instructions, see [the website][Diesel].
+
+These kinds of libraries are extremely powerful, but rely on custom derive for
+ergonomics. While these libraries _worked_ on Rust stable previously, they were
+not as nice to use, so much so that we often heard from users "I only use
+nightly because of Serde and Diesel." The use of custom derive is one of the
+most widely used nightly-only features. As such, [RFC 1681] was opened in July
+of last year to support this use-case. The RFC was merged in August, underwent
+a lot of development and testing, and now reaches stable today!
+
+[RFC 1681]: https://github.com/rust-lang/rfcs/pull/1681
+
+To find out how to write your own custom derives, see [the chapter of "The Rust
+Programming Language"](https://doc.rust-lang.org/book/procedural-macros.html).
+
+While we've said "Serde and Diesel" a number of times here, there's a lot of
+other cool things you can do with custom derive: see
+[`derive-new`](https://crates.io/crates/derive-new) for another example. Custom
+derive was also known as "macros 1.1", as it includes the infrastructure for
+supporting even more compile-time powers of Rust, nicknamed "macros 2.0."
+Expect to hear more about this space in future releases.
+
+As a smaller improvement, [`?Sized` can now be used in `where`
```

Comment:
You might put a section break here to visually indicate we're done talking about macros.